### ISSUE TITLE
Gazebo RFT dropping when any model is close to tall walls

### ISSUE THREAD

**Issue by SimoneMic** [icub-tech-iit/ergocub-software#65]
Hello

I am having an issue when spawning any robot close to a wall (or moving there after navigation) the RTF drops drastically (lower than 0.1).
The contacts are present on the whole world.
I have tried to remove the textures, but the effect is the same as in the following pictures:
**Warehouse example:**
![Screenshot from 2023-02-28 12-49-23](https://user-images.githubusercontent.com/86918431/221846698-fa2afe8b-a6c0-43af-a1d1-b8ca31eba7f4.png)

All the following situations have RFT smaller than 0.1

**Spawning iCub next to wall:**
![Screenshot from 2023-02-28 12-34-27](https://user-images.githubusercontent.com/86918431/221847056-88b4b62b-c8d5-4a0b-ac4b-21247e2f0da8.png)

**Spawning R1 next to wall:**
![Screenshot from 2023-02-28 12-33-23](https://user-images.githubusercontent.com/86918431/221847159-cfc95d6c-7211-47c8-956f-abbba919dcb3.png)

**Spawning ergoCubGazeboV1 next to wall:**
![Screenshot from 2023-02-28 12-14-02](https://user-images.githubusercontent.com/86918431/221847500-526e83b4-0761-4378-812f-0998c58aff52.png)

When spawning the robot in the middle of the scene instead, we have an higher RTF

**R1 in center of map RTF=0.99**
![Screenshot from 2023-02-28 12-38-32](https://user-images.githubusercontent.com/86918431/221848034-3a481310-550a-490e-9c76-1bdedbb887ea.png)

**ergoCubGazeboV1 in center of map RTF=0.6**
![Screenshot from 2023-02-28 12-39-27](https://user-images.githubusercontent.com/86918431/221848122-34607eb4-f84d-41dd-85ea-f00bc7a63276.png)

Does anybody have any suggestion on what could be the problem?
To me seems like a collision check threshold, that when the robot moves closer to the wall it starts checking for contacts.

I am also sharing my world file:
[worlds.zip](https://github.com/icub-tech-iit/ergocub-software/files/10849871/worlds.zip)

Thanks

cc @traversaro  @randaz81

**Comment by traversaro** [comment]
Interesting. Just to isolate the problem, if you comment out the laser the problem does not appear?

**Comment by traversaro** [comment]
(To double check, which version of Gazebo are you using?)

**Comment by randaz81** [comment]
> Interesting. Just to isolate the problem, if you comment out the laser the problem does not appear?

I'm also somehow convinced that is somehow connected to collision checking / lidar ray casting. The same issue was reported by kostantinos some time ago and never investigated in detail... ps: to prove that, i think that if you remove all collisions in the walls, the lidar rays will pass through the walls and maybe the simulation will run properly. Also, have gpu_ray plugin been tested?

**Comment by xEnVrE** [comment]
>  i think that if you remove all collisions in the walls, the lidar rays will pass through the walls and maybe the simulation will run properly.

I just tried that and indeed the ray pass through walls and RTF was back to the standard one. I am trying to simplify the collision meshes just to see what happens

**Comment by xEnVrE** [comment]
> Can you try if this happens also with other physics engine, in particular with dart (i.e. launch gazebo with -e dart)?

I think we tried once with @randaz81 and with DART the lidar plugin was not working as the ray casting routines are missing in DART

**Comment by traversaro** [comment]
> > Can you try if this happens also with other physics engine, in particular with dart (i.e. launch gazebo with -e dart)?
> 
> I think we tried once with @randaz81 and with DART the lidar plugin was not working as the ray casting routines are missing in DART If I recall correctly

Good call, related upstream issue: https://github.com/gazebosim/gazebo-classic/issues/911 . So the issue is related to ray, right? Then do you have any idea why "Spawning iCub next to wall:" also has a low RTF?

**Comment by traversaro** [comment]
So some pointers:
* What happens if you use `gpu_ray` as a sensor instead of `ray`?
* If you want to profile the code to understand what is going on, the possible way to proceed is : 
  * Profile what is going on with the Gazebo profiler ( https://classic.gazebosim.org/tutorials?tut=profiler&cat=tools_utilities ). Unfortunatly, this requires rebuilding Gazebo
  * Use an unmodified Gazebo binary with a profiler such as https://github.com/janestreet/magic-trace . However, I do not know if this will work fine as I do not know how much dbg information is available in the default Gazebo binary build.

**Comment by SimoneMic** [comment]
I did a separate test having `ergoCubGazeboV1` spawn outside the warehouse in a confined space using the wall builder inside Gazebo. The RTF is 0.98. 
The lidar rays are colliding with the walls.

![Screenshot from 2023-02-28 14-44-34](https://user-images.githubusercontent.com/86918431/221873017-93d09bc5-2471-413e-ab77-f3c96482127c.png)

![Screenshot from 2023-02-28 14-45-23](https://user-images.githubusercontent.com/86918431/221873054-165a6ad9-3022-48e7-8229-461e051184bc.png)

**Comment by SimoneMic** [comment]
Meanwhile, `stickBot` doesn't present this behavior.
RTF 0.88
![Screenshot from 2023-02-28 14-51-53](https://user-images.githubusercontent.com/86918431/221874119-aa239d5d-50ca-4ce3-86bb-b15e357e66ac.png)

The difference is that stickbot doesn't have collisions and meshes on its links.
![Screenshot from 2023-02-28 14-53-44](https://user-images.githubusercontent.com/86918431/221874533-e09555e6-a839-4635-a9eb-be364f61f620.png)
The only difference is that `stickBot` is on a previous docker without the newest gazebo yarp plugins (that are needed for `

I have tried to remove the collisions on `ergoCubGazeboV1` before, but I monitored the behavior only in the middle of the scene and there were not noticeable differences. Maybe in this extreme case will help, so I will try to redo it.

**Comment by SimoneMic** [comment]
I have removed the collisions on the urdf of `ergoCubGazeboV1` except for the `_sole` links (that don't use a mesh but a normal box from the sdf format)
RTF 0.75
![Screenshot from 2023-02-28 15-11-58](https://user-images.githubusercontent.com/86918431/221879716-d9982a8b-26f9-4380-a1b8-80a38a4870a9.png)

![Screenshot from 2023-02-28 15-12-43](https://user-images.githubusercontent.com/86918431/221879748-7f5df2fc-3fab-4e22-946b-b9f45fd80208.png)

Meanwhile, spawning the robot in the middle of the scenario, the improvements from the collisions are minimal. RTF 0.65 from 0.6
![Screenshot from 2023-02-28 15-15-48](https://user-images.githubusercontent.com/86918431/221880473-5fd29fa5-78e5-4d44-b11b-54f089ed7500.png)

**Comment by SimoneMic** [comment]
I think that the lession learned is to avoid using mesh files as collision surfaces, but only elemental geometrical volumes (like box, cylinder etc.).
Since the many mash element that are generated (I don't know how the segmentation of the finite element works in gazebo or by the tools used to generate the meshes and urdf in the repo) are checked by the collision monitor inside gazebo.

This explains the good behavior of `stickBot` that uses only elementary shapes. In fact I never had issues in simulation before.
I'll try to add some collision boxes on `ergoCubGazeboV1` (for the hands and torso, maybe) and see how it performs.

**Comment by traversaro** [comment]
@SimoneMic did you tried to see what happens if `gpu_ray` is used?

**Comment by SimoneMic** [comment]
I haven't tried using `gpu_ray` yet since it needs a bit of cherry-picking from the code that @morpheus1820 forked [gazebo-yarp-plugins](https://github.com/morpheus1820/gazebo-yarp-plugins/tree/gpu_ray/plugins/lasersensor) and update it to the newest release.
I am planning to do it tomorrow, since today I have managed to run a long and good simulation.

**Comment by SimoneMic** [comment]
I have tried using the `gpu_ray` but with no luck so far.
I have applyied the changes from [f0bc131](https://github.com/robotology/gazebo-yarp-plugins/compare/master...morpheus1820:gazebo-yarp-plugins:gpu_ray) to gazebo-yarp-plugins on my docker, and installed them.
In the urdf of the robot `ergoCubGazeboV1` I switched the type of sensor from `ray` to `gpu_ray` [here](https://github.com/icub-tech-iit/ergocub-software/blob/ac3f223dc2f183dea3f819369da0d58b59f1b2d3/urdf/ergoCub/robots/ergoCubGazeboV1/model.urdf#L3213) and [here](https://github.com/icub-tech-iit/ergocub-software/blob/ac3f223dc2f183dea3f819369da0d58b59f1b2d3/urdf/ergoCub/robots/ergoCubGazeboV1/model.urdf#L3245)

However, when I spawn the robot, the readings from the sensor are `Inf.` like the sensor is not finding any object.
Here the errors on gazebo:
`[ERROR] |gazebo-yarp-plugins.plugins.GazeboYarpLaserSensor| Size error, expected:  1600 , got:  0`

And a picture with the laser visualized:

![Screenshot from 2023-03-01 13-05-04](https://user-images.githubusercontent.com/86918431/222134654-0e897437-3a0d-4c57-a331-b4344320b9be.png)

I am launching my docker using this script:
```
#!/bin/bash
NAME=simonemiche/ergocub_ros2
TAG=ecub_first_year_demo
sudo xhost +
sudo docker run \
    --network=host --privileged \
    -it \
    --rm \
    -v /tmp/.X11-unix:/tmp/.X11-unix -e DISPLAY=unix${DISPLAY} --device /dev/dri/card0:/dev/dri/card0 --gpus=all \
    -e NVIDIA_DRIVER_CAPABILITIES=all --runtime=nvidia\
    ${NAME}:${TAG} bash
```
I don't know if there are any other files that need to be modified

**Comment by traversaro** [comment]
Thanks @SimoneMic ! It is indeed possible that there is a problem upstream in Gazebo.

**Comment by SimoneMic** [comment]
Closing since the "fix" will be discussed in #83
